### ISSUE TITLE
Add bindings support for setStyleJSON(const std::string&)

### ISSUE THREAD

**Issue by tobrun** [mapbox/mapbox-gl-native#6606]
In core in map.hpp/map.ccp we expose the followings method related to style loading:

```cpp
    void setStyleURL(const std::string&);
    void setStyleJSON(const std::string&);
```

In the Android bindings (and verified this is in macOS as well) we only support setting a style url, not setting a style JSON String. It seems that only the node bindings are implementing this currently and we are using this primarily in core for testing purposes.

Any history behind this that I'm missing here?

**Comment by 1ec5** [comment]
Yes, there’s a lot of history here: #900 #1299 #1367 #1997 #4769.

To recap: we originally did expose `setStyleJSON` in the iOS SDK, but I removed it per https://github.com/mapbox/mapbox-gl-native/issues/1147#issuecomment-88171503 and #900. I never implemented it in the macOS SDK.

The problem with `setStyleJSON()` is twofold: there are too many ways to populate a map view with content, so there are a lot of edge cases to test. Then there’s the issue of performance: parsing a giant JSON string is slow compared to all the other options.

There are two use cases for `setStyleJSON()`: either the developer is working with a large JSON document read in from a file, or they’re trying to manipulate a style on the fly. For the former case, we allow them to pass a local file URL into `setStyleURL()`. For the latter case, we have an object-oriented runtime styling API. We always envisioned this string-based API to be a stopgap in anticipation of the runtime styling API, anyways.

In spite of all that, I recently proposed something that would end up using `setStyleJSON()` under the hood in #6386: an initializer for MGLStyle that takes an NSData object holding JSON data. I mainly proposed this initializer for completeness’ sake, but unlike the old `-[MGLMapView setStyleJSON:]`, `-[MGLStyle initWithJSONData:]` would avoid adding complexity to MGLMapView. (The equivalent class on Android is perhaps `com.mapbox.mapboxsdk.constants.Style`, but it doesn’t represent a map view’s current style the way MGLStyle does on iOS and macOS.)
